### DISCUSSION TITLE
collection.load()

### DISCUSSION THREAD

**Discussion by chaochen1998** [milvus-io/milvus#33471] [Q&A and General discussion] [answered]
你好，我现在创建了一个包含20M个512维（类型为np.float16）向量的collection，在查询时，加载到内存用了37秒，查询的时间很短可以忽略不计，我想问一下：
1. 有什么办法可以加快这个速度吗？
2. 每进行一次单独的查询都要加载一次吗？可以只加载一次，然后在进行一系列查询后再释放吗？

**Comment by yhmo** [comment]
2000万个512维向量，数据量大约是38GB。load()的时间主要取决于数据量大小以及存储的读带宽。37秒读取38GB数据我觉得没什么问题，也没法加快。
collection.load()只需要调用一次，调用一次之后可以查询无数次。不用这个collection了就release掉，把内存空间释放。

**Reply by chaochen1998** [reply]
Q1

> collection.load()只需要调用一次，调用一次之后可以查询无数次。不用这个collection了就release掉，把内存空间释放。

可以理解为，我第一次执行某个查询脚本时将collection加载到cpu，这个脚本结束后我再一次运行这个脚本，同时将collection.load()语句注释掉，也是能work对吗？（目前我在我的设备上是work的）这中间的时间间隔也是没有限制的吗，在相关的docker一直处于运行状态的情况下。

Q2
我同时也在尝试milvus-lite，用的pymilvus版本为2.4.3，在运行例子的时候加入了创建partition的步骤出现了以下的错误：
```
# hello_milvus.py demonstrates the basic operations of PyMilvus, a Python SDK of Milvus.
# 1. connect to Milvus
# 2. create collection
# 3. insert data
# 4. create index
# 5. search, query, and hybrid search on entities
# 6. delete entities by PK
# 7. drop collection
import time

import numpy as np
from pymilvus import (
    connections,
    utility,
    FieldSchema, CollectionSchema, DataType,
    Collection,
)
from pymilvus.client.types import LoadState

fmt = "\n=== {:30} ===\n"
search_latency_fmt = "search latency = {:.4f}s"
num_entities, dim = 3000, 8

#################################################################################
# 1. connect to Milvus

print(fmt.format("start connecting to Milvus"))
connections.connect("default", uri="./local_test.db")
# connections.connect("default", host="localhost", port="19530")

has = utility.has_collection("hello_milvus")
print(f"Does collection hello_milvus exist in Milvus: {has}")

#################################################################################
# 2. create collection
# We're going to create a collection with 3 fields.
fields = [
    FieldSchema(name="random", dtype=DataType.DOUBLE),
    FieldSchema(name="pk", dtype=DataType.VARCHAR, is_primary=True, auto_id=False, max_length=100),
    FieldSchema(name="embeddings", dtype=DataType.FLOAT_VECTOR, dim=dim)
]

schema = CollectionSchema(fields, "hello_milvus is the simplest demo to introduce the APIs")

print(fmt.format("Create collection `hello_milvus`"))
hello_milvus = Collection("hello_milvus", schema, consistency_level="Strong")
# works ok
hello_milvus = Collection("hello_milvus", schema, consistency_level="Strong")

################################################################################
# 3. insert data
# We are going to insert 3000 rows of data into `hello_milvus`
# Data to be inserted must be organized in fields.
#
# The insert() method returns:
# - either automatically generated primary keys by Milvus if auto_id=True in the schema;
# - or the existing primary key field from the entities if auto_id=False in the schema.

hello_milvus.create_partition(partition_name="test")

print(fmt.format("Start inserting entities"))
rng = np.random.default_rng(seed=19530)
entities = [
    # provide the pk field because `auto_id` is set to False
    rng.random(num_entities).tolist(),  # field random, only supports list
    [str(i) for i in range(num_entities)],
    rng.random((num_entities, dim)),    # field embeddings, supports numpy.ndarray and list
]

insert_result = hello_milvus.insert(entities,partition_name="test")

```
报错为：
```

=== start connecting to Milvus     ===

Does collection hello_milvus exist in Milvus: True

=== Create collection `hello_milvus` ===

grpc RpcError: [has_partition], <_MultiThreadedRendezvous: StatusCode.UNIMPLEMENTED, >, <Time:{'RPC start': '2024-05-30 05:40:29.346578', 'gRPC error': '2024-05-30 05:40:29.346929'}>
Traceback (most recent call last):
  File "/vector_db/test.py", line 74, in <module>
    hello_milvus.create_partition(partition_name="test")
  File "/opt/conda/lib/python3.10/site-packages/pymilvus/orm/collection.py", line 1188, in create_partition
    if self.has_partition(partition_name, **kwargs) is True:
  File "/opt/conda/lib/python3.10/site-packages/pymilvus/orm/collection.py", line 1222, in has_partition
    return conn.has_partition(self._name, partition_name, timeout=timeout, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/pymilvus/decorators.py", line 161, in handler
    raise e from e
  File "/opt/conda/lib/python3.10/site-packages/pymilvus/decorators.py", line 143, in handler
    return func(*args, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/pymilvus/decorators.py", line 182, in handler
    return func(self, *args, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/pymilvus/decorators.py", line 91, in handler
    raise e from e
  File "/opt/conda/lib/python3.10/site-packages/pymilvus/decorators.py", line 87, in handler
    return func(*args, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/pymilvus/client/grpc_handler.py", line 428, in has_partition
    response = rf.result()
  File "/opt/conda/lib/python3.10/site-packages/grpc/_channel.py", line 883, in result
    raise self
grpc._channel._MultiThreadedRendezvous: <_MultiThreadedRendezvous of RPC that terminated with:
        status = StatusCode.UNIMPLEMENTED
        details = ""
        debug_error_string = "UNKNOWN:Error received from peer  {created_time:"2024-05-30T05:40:29.346762813+00:00", grpc_status:12, grpc_message:""}"
>
```
貌似milvus-lite无法创建partiton吗？

**Reply by yhmo** [reply, marked as answer]
load()的意思是：通知milvus server，把这个collection的数据加载到服务端的内存里，只要你不用collection.release()去通知释放，那些数据就一直保持在服务端内存里。只有数据在内存里才能被查询。所以只要你调用了一次collection.load()，只要你不重启milvus服务端，那随便你客户端调多少次查询都行，调用了一次collection.load()之后就不需要再次调用load()，实际上就算你调用多次collection.load()，也只有第一次干活。

**Reply by chaochen1998** [reply]
明白了，感谢回复。
另外，第二个问题您了解吗？

**Reply by yhmo** [reply]
目前milvus-lite暂时还不支持partition的功能，所以报错是UNIMPLEMENTED

**Reply by chaochen1998** [reply]
明白了，再次感谢回复

**Reply by cutelitchi** [reply]
假设内存大小一定，如果我持续加载多个collection，理论上肯定会到某一个加载得时候提示加载失败，我想做的是，定期清理调用时间最远的一些collection，但是现在遇到一个问题就是，queryNode初始化时，未加载collection，占用内存大概再200-300MB左右，加载了一个1GB左右的collection后，又release掉，但是内存并没有释放很多，而是保留在700-800MB左右，请问这个release操作，到底可以释放多少空间？感谢
图1：初始化占用内存
图2：load一个1GBcollection后，又release掉后的内存占用
![init](https://github.com/user-attachments/assets/112f9640-daa3-476b-a9bb-178c16627d68)
![释放后](https://github.com/user-attachments/assets/2f1db92d-6d51-418f-816a-40bf7ec51b15)

**Reply by xiaofan-luan** [reply]
> 假设内存大小一定，如果我持续加载多个collection，理论上肯定会到某一个加载得时候提示加载失败，我想做的是，定期清理调用时间最远的一些collection，但是现在遇到一个问题就是，queryNode初始化时，未加载collection，占用内存大概再200-300MB左右，加载了一个1GB左右的collection后，又release掉，但是内存并没有释放很多，而是保留在700-800MB左右，请问这个release操作，到底可以释放多少空间？感谢 图1：初始化占用内存 图2：load一个1GBcollection后，又release掉后的内存占用 ![init](https://private-user-images.githubusercontent.com/9426798/372025360-112f9640-daa3-476b-a9bb-178c16627d68.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc3NDkzNTgsIm5iZiI6MTcyNzc0OTA1OCwicGF0aCI6Ii85NDI2Nzk4LzM3MjAyNTM2MC0xMTJmOTY0MC1kYWEzLTQ3NmItYTliYi0xNzhjMTY2MjdkNjgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAwMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMDFUMDIxNzM4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZDA5NGNmMTE5MjFiM2E4YjBjNjQwZGQ0ZTFlMDgyZTUyOTJlNzUwYjU4NGQ0MTkwOTdlOTBjYTFjMmUyNTIxMSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.hiI-h1Wsn2gAMqQceP_CCGyKOUujC8d96UEcZWsaEgo) ![释放后](https://private-user-images.githubusercontent.com/9426798/372025728-2f1db92d-6d51-418f-816a-40bf7ec51b15.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc3NDkzNTgsIm5iZiI6MTcyNzc0OTA1OCwicGF0aCI6Ii85NDI2Nzk4LzM3MjAyNTcyOC0yZjFkYjkyZC02ZDUxLTQxOGYtODE2YS00MGJmN2VjNTFiMTUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAwMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMDFUMDIxNzM4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YzYxZDcxNDY0NTVlNWQzYzU0OWY4ZjhjZTU0MmViYWEyOGY3NDJmMDRkOGZmNDkwMWZlZGNhYWFhYjkyNGYzNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.nDYvVbS9yMrbTSUJEZ08y5qPUppFXVqR5ndHvTgUKSk)

这部分内存实际上被内存分配器缓存了。所以只要持续加载即可。如果你希望看到明显的内存下降，可能需要一些内存分配器的参数，比如
export MALLOC_CONF=dirty_decay_ms:1000,muzzy_decay_ms:1000
事实上，当内存比较紧张的时候，应该会看到内存分配器自动回收。

从S3上加载数据本身比较慢，不适合在线访问，只能作为离线查询（延迟s级别以上）

如果你对延迟要求不高，几秒就可以，可以试试LRU模式，否则建议使用MMap模式（需要有高性能的NVMe磁盘）。mmap模式可以实现百ms级别的查询延迟

**Reply by cutelitchi** [reply]
明白了，非常感谢解答！👍
